Code/Framework/AzToolsFramework/AzToolsFramework/Entity/PrefabEditorEntityOwnershipService.cpp

```diff
@@ -175,6 +180,11 @@ namespace AzToolsFramework
         AZ_Assert(IsInitialized(), "Tried to destroy an entity without initializing the Entity Ownership Service");
         AZ_Assert(m_entitiesRemovedCallback, "Callback function for DestroyEntityById has not been set.");
         OnEntityRemoved(entityId);
+        auto owningInstance = m_instanceEntityMapperInterface->FindOwningInstance(entityId);
+        if (owningInstance.has_value())
+        {
+            owningInstance->get().DetachEntity(entityId).release();
```

Comment:
If a system tries to call the EOS to  DestroyEntityById(), this function now has the likelihood of creating a memory leak if the caller is not aware that they need to destroy the entity themselves.

Rather than doing it this way,
~Entity()->OnEntityDestruction()->DetachEntity(entityId).release() might be a better way to handle this.

This would require 
1. Adding an OnEntityDestroyed() call to EntityOwnershipService.h
2. Making the SliceEOS override it to call DestroyEntityById() (no change here)
3. Making the PrefabEditorEOS override it to detach and release.

Ideally, I don't think other systems should be managing the lifecycle of entities because it should be the prefab system that does it. That's why prefab instances have AZStd::unique_ptr<AZ::Entity> in them. But access to raw pointer to entities has been given prior to prefabs and exists for legacy reasons.

If anyone from physx is reading this, I recommend that they change their unit tests implementation to not manage entity life cycles themselves.

Reply:
Even if entity deletion was managed by the prefab system, DestroyEntity still wouldn't have detached the entity from the Instance. Which call would a user make if they wanted to delete an entity?

Reply:
@michabr, can you create a github issue asking physx team to refactor their tests to not manage entity lifecycles themselves?

Reply:
It looks like DestroyEntityById and DestroyEntity have different purposes which is a bit confusing. DestroyEntityById is called via ~Entity() and shouldn't delete the entity since it's already being deleted. DestroyEntity is an API called by the user and so the function should delete the entity after detaching it. I'll double check that this is happening.

Reply:
> Even if entity deletion was managed by the prefab system, DestroyEntity still wouldn't have detached the entity from the Instance. Which call would a user make if they wanted to delete an entity?

They can call ToolsApplication::DeleteEntityById() or EditorEntityManager::DeleteEntityById or  PrefabPublicHandler::DeleteEntitiesAndAllDescendantsInInstance() which would correctly detach and also update corresponding prefab templates. 

> DestroyEntityById is called via ~Entity() and shouldn't delete the entity since it's already being deleted. 

yes but I don't think that's the function we should be calling. DesroyEntityById() is called from EntityContext::OnEntityDestruction(). Instead of doing that, I am recommending the introducton of a new EOS function to avoid confusion that we won't be actually destroying the entity but rather let the EOS do cleanup for the destroyed entity. Feel free to reach out to me if you have questions on this approach.

Reply:
I wasn't familiar with the EditorEntityManager but it looks like that is a wrapper around prefab public handler? Would it be correct to say that EditorEntityManager/PrefabPublicHandler is for calls that also handle undo/redo nodes for you, and direct API calls to EOS should not do any undo/redo processing?

Since the EOS also has an API to destroy an entity, I think we should still handle that correctly and actually detach and delete the entity when called.

Reply:
Made DestroyEntity and DestroyEntityById behave the same (detaches and deletes the entity), and added a new method to replace the DestroyEntityById in EntityContext::OnEntityDestruction() that only detaches the entity from its instance.

Reply:
> Would it be correct to say that EditorEntityManager/PrefabPublicHandler is for calls that also handle undo/redo nodes for you, and direct API calls to EOS should not do any undo/redo processing

That's a fair callout but one thing to remember is that slices uses EOS for both editor context and game context. So, being able to do something without undo/redo support makes sense. But prefabs only have an Editor EOS and prefabs are only an editor concept. So I don't think providing entity operations without undo/redo support for prefabs is something that we should be supporting. I can't think of an actual real life editor use case for being able to modify prefabs without undo/redo support. Let's say for example, you remove an entity from a prefab instance, what use is that prefab instance if it is not in sync with the underlying prefab template?

Also, PrefabEditorEOS::AddEntity() and PrefabEditorEOS::AddEntities() do support undo/redo. So, right now there doesn't appear to be a streamlined approach defined to say what the PrefabEditorEOS should be doing. In the long run, we should try to unify PrefabEditorEOS and PrefabPublicHandler into one api because they both serve the same purpose- provide a way to modify editor entities while keeping the underlying in-memory prefabs in the correct states.

Overall, I think it is okay to add support for DestroyEntityById() because it satisfies the EOS requirements but we should really streamline the EOS and PublicHandler APIs and decide how they are going to co-exist going forward. @michabr , can you create a github issue to track this work?

Reply:
Curious: Is release() the right call to destroy the entity object?

Ref: https://stackoverflow.com/questions/25609457/does-unique-ptrrelease-call-the-destructor

Reply:
No release explicitly means that the unique pointer gives up owernship of it's address.
For destruction, `reset()` would need to be called

Reply:
Here, we don't want to destroy the entity object.